File: packages/Thoth.Json.JavaScript/Encode.fs
namespace Thoth.Json.JavaScript

open Fable.Core
open Fable.Core.JsInterop
open Thoth.Json.Core
open System.Globalization

[<RequireQualifiedAccess>]
module Encode =

    /// <summary>
    /// Builds a <c>obj</c>, so an encoder written against Thoth.Json.Core runs on JavaScript.
    /// </summary>
    let helpers =
        { new IEncoderHelpers<obj> with
            member _.encodeString value = box value
            member _.encodeChar value = box value
            member _.encodeDecimalNumber value = box value
            member _.encodeBool value = box value
            member _.encodeNull() = box null

            member _.encodeObject(values) =
                let o = obj ()

                for key, value in values do
                    o?(key) <- value

                o

            member _.encodeArray values = JS.Constructors.Array.from values
            member _.encodeList values = JS.Constructors.Array.from values
            member _.encodeSeq values = JS.Constructors.Array.from values

            member _.encodeResizeArray values =
                JS.Constructors.Array.from values

            member _.encodeSignedIntegralNumber value = box value
            member _.encodeUnsignedIntegralNumber value = box value
        }

    /// <summary>
    /// Write an encodable as a JSON string, indented by the given number of spaces. Pass <c>0</c>
    /// for a single line.
    /// </summary>
    let toString (space: int) (value: IEncodable) : string =
        let json = Encode.toJsonValue helpers value
        JS.JSON.stringify (json, space = space)

    /// <summary>
    /// Write a value as a JSON string with the encoder half of a codec.
    /// </summary>
    let fromCodec (space: int) (codec: Codec<'T>) (value: 'T) : string =
        value |> Encode.codec codec |> toString space

File: assets/live-example/precompiled/18264c7b6c523f27/packages/Thoth.Json.JavaScript/Encode.js

import { disposeSafe, getEnumerator, defaultOf } from "fable-library-js/Util.js";
import { codec as codec_1 } from "../Thoth.Json.Core/Encode.js";

export const helpers = {
    encodeString(value) {
        return value;
    },
    encodeChar(value_1) {
        return value_1;
    },
    encodeDecimalNumber(value_2) {
        return value_2;
    },
    encodeBool(value_3) {
        return value_3;
    },
    encodeNull() {
        return defaultOf();
    },
    encodeObject(values) {
        const o = {};
        const enumerator = getEnumerator(values);
        try {
            while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
                const forLoopVar = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
                o[forLoopVar[0]] = forLoopVar[1];
            }
        }
        finally {
            disposeSafe(enumerator);
        }
        return o;
    },
    encodeArray(values_1) {
        return Array.from(values_1);
    },
    encodeList(values_2) {
        return Array.from(values_2);
    },
    encodeSeq(values_3) {
        return Array.from(values_3);
    },
    encodeResizeArray(values_4) {
        return Array.from(values_4);
    },
    encodeSignedIntegralNumber(value_5) {
        return value_5;
    },
    encodeUnsignedIntegralNumber(value_6) {
        return value_6;
    },
};

/**
 * Write an encodable as a JSON string, indented by the given number of spaces. Pass <c>0</c>
 * for a single line.
 */
export function toString(space, value) {
    const json_1 = value.Encode(helpers);
    return JSON.stringify(json_1, undefined, space);
}

/**
 * Write a value as a JSON string with the encoder half of a codec.
 */
export function fromCodec(space, codec, value) {
    return toString(space, codec_1(codec)(value));
}

